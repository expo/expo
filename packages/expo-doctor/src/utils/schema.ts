// copied from https://github.com/expo/expo-cli/blob/d00319aae4fdcacf1a335af5a8428c45b62fc4d7/packages/xdl/src/project/Doctor.ts
// minor naming changes only

import { ExpoConfig } from '@expo/config';
import Schemer, { SchemerError, ValidationError } from '@expo/schemer';

import { learnMore } from './TerminalLink';

function formatValidationError(validationError: ValidationError) {
  return `\n â€¢ ${validationError.fieldPath ? 'Field: ' + validationError.fieldPath + ' - ' : ''}${
    validationError.message
  }.`;
}

export async function validateWithSchemaAsync(
  projectRoot: string,
  {
    // Extract internal from the config object.
    _internal,
    ...exp
  }: ExpoConfig,
  schema: any,
  configName: string,
  validateAssets: boolean
): Promise<{ schemaErrorMessage: string | undefined; assetsErrorMessage: string | undefined }> {
  let schemaErrorMessage;
  let assetsErrorMessage;
  const validator = new Schemer(schema, { rootDir: projectRoot });

  // Validate the schema itself
  try {
    await validator.validateSchemaAsync(exp);
  } catch (e: any) {
    if (e instanceof SchemerError) {
      schemaErrorMessage = `Error: Problem${
        e.errors.length > 1 ? 's' : ''
      } validating fields in ${configName}. ${learnMore(
        'https://docs.expo.dev/workflow/configuration/'
      )}`;
      schemaErrorMessage += e.errors.map(formatValidationError).join('');
    }
  }

  if (validateAssets) {
    try {
      await validator.validateAssetsAsync(exp);
    } catch (e: any) {
      if (e instanceof SchemerError) {
        assetsErrorMessage = `Error: Problem${
          e.errors.length > 1 ? '' : 's'
        } validating asset fields in ${configName}. ${learnMore('https://docs.expo.dev/')}`;
        assetsErrorMessage += e.errors.map(formatValidationError).join('');
      }
    }
  }
  return { schemaErrorMessage, assetsErrorMessage };
}
