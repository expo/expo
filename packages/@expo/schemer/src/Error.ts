export class SchemerError extends Error {
  readonly name = 'SchemerError';
  errors: ValidationError[];

  constructor(errors: ValidationError[]) {
    super('');
    this.message = errors.map((error) => error.message).join('\n');
    this.errors = errors;
  }
}

export class ValidationError extends Error {
  readonly name = 'ValidationError';
  errorCode: string;
  fieldPath: string;
  data: any;
  meta: any;
  constructor({
    errorCode,
    fieldPath,
    message,
    data,
    meta,
  }: {
    errorCode: ErrorCode;
    fieldPath: string;
    message: string;
    data: any;
    meta: any;
  }) {
    super(message);
    this.errorCode = errorCode;
    this.fieldPath = fieldPath;
    this.data = data;
    this.meta = meta;
  }
}

export type ErrorCode = keyof typeof ErrorCodes;

export const ErrorCodes = {
  SCHEMA_VALIDATION_ERROR: 'SCHEMA_VALIDATION_ERROR',
  SCHEMA_ADDITIONAL_PROPERTY: 'SCHEMA_ADDITIONAL_PROPERTY',
  SCHEMA_MISSING_REQUIRED_PROPERTY: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
  SCHEMA_INVALID_PATTERN: 'SCHEMA_INVALID_PATTERN',
  SCHEMA_INVALID_NOT: 'SCHEMA_INVALID_NOT',

  INVALID_ASSET_URI: 'INVALID_ASSET_URI',
  INVALID_DIMENSIONS: 'INVALID_DIMENSIONS',
  INVALID_CONTENT_TYPE: 'INVALID_CONTENT_TYPE',
  NOT_SQUARE: 'NOT_SQUARE',
  FILE_EXTENSION_MISMATCH: 'FILE_EXTENSION_MISMATCH',
};
