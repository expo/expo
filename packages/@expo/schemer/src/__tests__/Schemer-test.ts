import good from './fixtures/app.json';
import bad from './fixtures/bad.json';
import badWithNot from './fixtures/badwithnot.json';
import invalidAppIcon from './fixtures/invalidAppIcon.json';
import schema from './fixtures/schema.json';
import Schemer from '..';
import { ErrorCodes, SchemerError } from '../Error';

const validator = new Schemer(schema.schema, { rootDir: __dirname });

describe('Sanity Tests', () => {
  it('returns instance of Schemer', () => {
    expect(validator instanceof Schemer).toBe(true);
  });

  it('returns instance with public functions', () => {
    expect(validator).toMatchObject({
      validateAll: expect.any(Function),
      validateProperty: expect.any(Function),
    });
  });
});

describe('Image Validation', () => {
  it('errors for webp images', async () => {
    const error = await expectSchemerToThrowAsync(() =>
      validator.validateAssetsAsync({
        android: {
          adaptiveIcon: { foregroundImage: './fixtures/webp.webp' },
        },
      })
    );

    expect(error.errors).toEqual([
      expect.objectContaining({ errorCode: ErrorCodes.INVALID_CONTENT_TYPE }),
      expect.objectContaining({ errorCode: ErrorCodes.NOT_SQUARE }),
    ]);
    expectSchemerErrorToMatchSnapshot(error);
  });

  it('errors when file extension and content do not match up', async () => {
    const error = await expectSchemerToThrowAsync(() =>
      validator.validateAssetsAsync({
        icon: './fixtures/secretlyPng.jpg',
      })
    );

    expect(error.errors).toEqual([
      expect.objectContaining({ errorCode: ErrorCodes.FILE_EXTENSION_MISMATCH }),
    ]);
    expectSchemerErrorToMatchSnapshot(error);
  });
});

describe('Holistic Unit Test', () => {
  it('good example app.json all', async () => {
    expect(await validator.validateAll(good)).toBeUndefined();
  });

  it('good example app.json schema', async () => {
    expect(await validator.validateSchemaAsync(good)).toBeUndefined();
  });

  it('bad example app.json schema', async () => {
    expectSchemerErrorToMatchSnapshot(
      await expectSchemerToThrowAsync(() => validator.validateSchemaAsync(bad))
    );
  });

  it('bad example app.json schema with field with not', async () => {
    expectSchemerErrorToMatchSnapshot(
      await expectSchemerToThrowAsync(() => validator.validateSchemaAsync(badWithNot))
    );
  });

  it('bad example app.json - invalid path for app icon', async () => {
    expectSchemerErrorToMatchSnapshot(
      await expectSchemerToThrowAsync(() => validator.validateAll(invalidAppIcon))
    );
  });
});

describe('Manual Validation Individual Unit Tests', () => {
  it('Local Icon', async () => {
    expect(await validator.validateIcon('./fixtures/check.png')).toBeUndefined();
  });

  it('Local Square Icon correct', async () => {
    const customValidator = new Schemer(
      { properties: { icon: { meta: { asset: true, square: true } } } },
      { rootDir: __dirname }
    );
    expect(await customValidator.validateIcon('./fixtures/check.png')).toBeUndefined();
  });

  it('Local icon dimensions wrong', async () => {
    const customValidator = new Schemer(
      {
        properties: {
          icon: {
            meta: {
              asset: true,
              dimensions: { width: 400, height: 401 },
              contentTypePattern: '^image/png$',
            },
          },
        },
      },
      { rootDir: __dirname }
    );
    const error = await expectSchemerToThrowAsync(() =>
      customValidator.validateIcon('./fixtures/check.png')
    );

    expect(error.errors).toHaveLength(1);
    expectSchemerErrorToMatchSnapshot(error);
  });
});

describe('Individual Unit Tests', () => {
  it('Error when missing Required Property', async () => {
    const customValidator = new Schemer(
      {
        properties: { name: {} },
        required: ['name'],
      },
      { rootDir: __dirname }
    );
    const error = await expectSchemerToThrowAsync(() =>
      customValidator.validateAll({ noName: '' })
    );

    expect(error.errors).toHaveLength(1);
    expect(error.errors).toEqual([
      expect.objectContaining({ errorCode: ErrorCodes.SCHEMA_MISSING_REQUIRED_PROPERTY }),
    ]);
    expectSchemerErrorToMatchSnapshot(error);
  });

  it('Error when data has an additional property', async () => {
    const customValidator = new Schemer({ additionalProperties: false }, { rootDir: __dirname });

    const error = await expectSchemerToThrowAsync(() =>
      customValidator.validateAll({ extraProperty: 'extra' })
    );

    expect(error.errors).toHaveLength(1);
    expect(error.errors).toEqual([
      expect.objectContaining({ errorCode: ErrorCodes.SCHEMA_ADDITIONAL_PROPERTY }),
    ]);
    expectSchemerErrorToMatchSnapshot(error);
  });

  it.each`
    name            | expectedError
    ${'wilson'}     | ${undefined}
    ${[1, 2, 3, 4]} | ${'must be string'}
    ${23.232332}    | ${'must be string'}
    ${/regex.*/}    | ${'must be string'}
  `('validates name: $name', async ({ name, expectedError }) => {
    if (!expectedError) {
      expect(await validator.validateName(name)).toBeUndefined();
    } else {
      const error = await expectSchemerToThrowAsync(() => validator.validateName(name));
      expect(error.message).toBe(expectedError);
    }
  });

  it.each`
    slug                                             | expectedError
    ${'wilson'}                                      | ${undefined}
    ${12312123123}                                   | ${'must be string'}
    ${[1, 23]}                                       | ${'must be string'}
    ${'wilson123'}                                   | ${undefined}
    ${'wilson-123'}                                  | ${undefined}
    ${'wilson/test'}                                 | ${'\'\' must match pattern "^[a-zA-Z0-9_\\-]+$"'}
    ${'wilson-test%'}                                | ${'\'\' must match pattern "^[a-zA-Z0-9_\\-]+$"'}
    ${'wilson-test-zhao--javascript-is-super-funky'} | ${undefined}
  `('validates slug: $slug', async ({ slug, expectedError }) => {
    if (!expectedError) {
      expect(await validator.validateSlug(slug)).toBeUndefined();
    } else {
      const error = await expectSchemerToThrowAsync(() => validator.validateSlug(slug));
      expect(error.message).toBe(expectedError);
    }
  });

  it.each`
    sdkVersion       | expectedError
    ${'1.0.0'}       | ${undefined}
    ${'2.0.0.0.1'}   | ${undefined}
    ${'UNVERSIONED'} | ${undefined}
    ${'12.2a.3'}     | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
    ${'9,9,9'}       | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
    ${'1.2'}         | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
  `('validates SDK version: $sdkVersion', async ({ sdkVersion, expectedError }) => {
    if (!expectedError) {
      expect(await validator.validateSdkVersion(sdkVersion)).toBeUndefined();
    } else {
      const error = await expectSchemerToThrowAsync(() => validator.validateSdkVersion(sdkVersion));
      expect(error.message).toBe(expectedError);
    }
  });
});

async function expectSchemerToThrowAsync(action: () => any): Promise<SchemerError> {
  try {
    await action();
  } catch (error: any) {
    expect(error).toBeInstanceOf(SchemerError);
    return error;
  }

  throw new Error('Expression did not throw the expected error');
}

function expectSchemerErrorToMatchSnapshot(error: SchemerError) {
  expect(
    error.errors.map((validationError) => {
      const { stack, message, ...rest } = validationError;
      return { ...rest, message };
    })
  ).toMatchSnapshot();
}
