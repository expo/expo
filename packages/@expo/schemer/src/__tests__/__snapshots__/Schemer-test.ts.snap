// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holistic Unit Test bad example app.json - invalid path for app icon 1`] = `
[
  {
    "data": "./unknown/path.png",
    "errorCode": "INVALID_ASSET_URI",
    "fieldPath": "icon",
    "message": "cannot access file at './unknown/path.png'",
    "meta": {
      "asset": true,
      "bareWorkflow": "To change your app's icon, edit or replace the files in \`ios/<PROJECT-NAME>/Assets.xcassets/AppIcon.appiconset\` (we recommend using Xcode), and \`android/app/src/main/res/mipmap-<RESOLUTION>\`. Be sure to follow the guidelines for each platform ([iOS](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/app-icon/), [Android 7.1 and below](https://material.io/design/iconography/#icon-treatments), and [Android 8+](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive)) and to provide your new icon in each existing size.",
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Holistic Unit Test bad example app.json schema 1`] = `
[
  {
    "data": {
      "asdfasdfandroid": {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_MISSING_REQUIRED_PROPERTY",
    "fieldPath": "",
    "message": "is missing required property 'name'",
    "meta": undefined,
    "name": "ValidationError",
  },
  {
    "data": {
      "asdfasdfandroid": {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'orientaasdfasdftion'",
    "meta": undefined,
    "name": "ValidationError",
  },
  {
    "data": {
      "asdfasdfandroid": {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'asdfasdfandroid'",
    "meta": undefined,
    "name": "ValidationError",
  },
  {
    "data": "1*@)#($*@)(#$*)",
    "errorCode": "SCHEMA_INVALID_PATTERN",
    "fieldPath": "slug",
    "message": "'slug' must match pattern \"^[a-zA-Z0-9_\\-]+$\"",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Holistic Unit Test bad example app.json schema with field with not 1`] = `
[
  {
    "data": "1.0",
    "errorCode": "SCHEMA_INVALID_NOT",
    "fieldPath": "runtimeVersion",
    "message": "'runtimeVersion' should be not a decimal ending in a 0.",
    "meta": {
      "notHuman": "Not a decimal ending in a 0.",
    },
    "name": "ValidationError",
  },
]
`;

exports[`Image Validation errors for webp images 1`] = `
[
  {
    "data": "./fixtures/webp.webp",
    "errorCode": "INVALID_CONTENT_TYPE",
    "fieldPath": "Android.adaptiveIcon.foregroundImage",
    "message": "field 'Android.adaptiveIcon.foregroundImage' should point to .png image but the file at './fixtures/webp.webp' has type webp",
    "meta": {
      "asset": true,
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
  {
    "data": "./fixtures/webp.webp",
    "errorCode": "NOT_SQUARE",
    "fieldPath": "Android.adaptiveIcon.foregroundImage",
    "message": "image should be square, but the file at './fixtures/webp.webp' has dimensions 320x214",
    "meta": {
      "asset": true,
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Image Validation errors when file extension and content do not match up 1`] = `
[
  {
    "data": "./fixtures/secretlyPng.jpg",
    "errorCode": "FILE_EXTENSION_MISMATCH",
    "fieldPath": "icon",
    "message": "the file extension should match the content, but the file extension is .jpg while the file content at './fixtures/secretlyPng.jpg' is of type png",
    "meta": {
      "asset": true,
      "bareWorkflow": "To change your app's icon, edit or replace the files in \`ios/<PROJECT-NAME>/Assets.xcassets/AppIcon.appiconset\` (we recommend using Xcode), and \`android/app/src/main/res/mipmap-<RESOLUTION>\`. Be sure to follow the guidelines for each platform ([iOS](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/app-icon/), [Android 7.1 and below](https://material.io/design/iconography/#icon-treatments), and [Android 8+](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive)) and to provide your new icon in each existing size.",
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Individual Unit Tests Error when data has an additional property 1`] = `
[
  {
    "data": {
      "extraProperty": "extra",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'extraProperty'",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Individual Unit Tests Error when missing Required Property 1`] = `
[
  {
    "data": {
      "noName": "",
    },
    "errorCode": "SCHEMA_MISSING_REQUIRED_PROPERTY",
    "fieldPath": "",
    "message": "is missing required property 'name'",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Manual Validation Individual Unit Tests Local icon dimensions wrong 1`] = `
[
  {
    "data": "./fixtures/check.png",
    "errorCode": "INVALID_DIMENSIONS",
    "fieldPath": "icon",
    "message": "'icon' should have dimensions 400x401, but the file at './fixtures/check.png' has dimensions 512x512",
    "meta": {
      "asset": true,
      "contentTypePattern": "^image/png$",
      "dimensions": {
        "height": 401,
        "width": 400,
      },
    },
    "name": "ValidationError",
  },
]
`;
